tests/cpp_unit_tests/test_math.cpp

```diff
@@ -0,0 +1,364 @@
+  std::vector<double> ref_vals;
+
+  for (int n = 0; n <= max_order; ++n) {
+    for (int m = -n; m <= n; ++m) {
+      double ylm;
+      if (m < 0) {
+        ylm = pnm_bar(n, std::abs(m), mu) * std::sin(std::abs(m) * azi);
+      } else {
+        ylm = pnm_bar(n, m, mu) * std::cos(m * azi);
+      }
+
+      // Un-normalize for comparison
+      ylm /= std::sqrt(2.0 * n + 1);
+      ref_vals.push_back(ylm);
+    }
+  }
+
+  std::vector<double> test_uvw {std::sin(pol) * std::cos(azi),
+    std::sin(pol) * std::sin(azi), std::cos(pol)};
+
+  std::vector<double> test_vals((max_order + 1) * (max_order + 1), 0);
+  openmc::calc_rn_c(max_order, test_uvw.data(), test_vals.data());
+
+  REQUIRE_THAT(ref_vals, Catch::Matchers::Approx(test_vals));
+}
+
+TEST_CASE("Test calc_zn")
+{
+  int n = 10;
+  double rho = 0.5;
+  double phi = 0.5;
+
+  std::vector<double> ref_vals {1.00000000e+00, 2.39712769e-01, 4.38791281e-01,
+    2.10367746e-01, -5.00000000e-01, 1.35075576e-01, 1.24686873e-01,
+    -2.99640962e-01, -5.48489101e-01, 8.84215021e-03, 5.68310892e-02,
+    -4.20735492e-01, -1.25000000e-01, -2.70151153e-01, -2.60091773e-02,
+    1.87022545e-02, -3.42888902e-01, 1.49820481e-01, 2.74244551e-01,
+    -2.43159131e-02, -2.50357380e-02, 2.20500013e-03, -1.98908812e-01,
+    4.07587508e-01, 4.37500000e-01, 2.61708929e-01, 9.10321205e-02,
+    -1.54686328e-02, -2.74049397e-03, -7.94845816e-02, 4.75368705e-01,
+    7.11647284e-02, 1.30266162e-01, 3.37106977e-02, 1.06401886e-01,
+    -7.31606787e-03, -2.95625975e-03, -1.10250006e-02, 3.55194307e-01,
+    -1.44627826e-01, -2.89062500e-01, -9.28644588e-02, -1.62557358e-01,
+    7.73431638e-02, -2.55329539e-03, -1.90923851e-03, 1.57578403e-02,
+    1.72995854e-01, -3.66267690e-01, -1.81657333e-01, -3.32521518e-01,
+    -2.59738162e-02, -2.31580576e-01, 4.20673902e-02, -4.11710546e-04,
+    -9.36449487e-04, 1.92156884e-02, 2.82515641e-02, -3.90713738e-01,
+    -1.69280296e-01, -8.98437500e-02, -1.08693628e-01, 1.78813094e-01,
+    -1.98191857e-01, 1.65964201e-02, 2.77013853e-04};
+
+  int nums = ((n + 1) * (n + 2)) / 2;
+
+  std::vector<double> test_vals(nums, 0);
+  openmc::calc_zn(n, rho, phi, test_vals.data());
+
+  REQUIRE_THAT(ref_vals, Catch::Matchers::Approx(test_vals));
+}
+
+TEST_CASE("Test calc_zn_rad")
+{
+  int n = 10;
+  double rho = 0.5;
+
+  std::vector<double> ref_vals {1.00000000e+00, -5.00000000e-01,
+    -1.25000000e-01, 4.37500000e-01, -2.89062500e-01, -8.98437500e-02};
+
+  int nums = n / 2 + 1;
+  std::vector<double> test_vals(nums, 0);
+  openmc::calc_zn_rad(n, rho, test_vals.data());
+
+  REQUIRE_THAT(ref_vals, Catch::Matchers::Approx(test_vals));
+}
+
+TEST_CASE("Test rotate_angle")
+{
+  std::vector<double> uvw0 {1.0, 0.0, 0.0};
+  double phi = 0.0;
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_int_distribution<> dist(0, 1);
+
+  uint64_t prn_seed = 0;
+  for (int i = 0; i < 63; ++i) {
+    prn_seed |= static_cast<std::uint64_t>(dist(gen)) << i;
```

Comment:
What is this code doing?

Reply:
emm, I see that in python api

```
if prn_seed is None:
    prn_seed = getrandbits(63)
```
It generated a random integer of k bits length, so I just want to imitate the behavior, but as you mentioned, maybe I can change it use OpenMC's random number generator here

Reply:
Since it's only for testing, this should work just fine.